src/utils.jl

```diff
@@ -243,17 +243,442 @@ function generate_nonallocating_forces(calc_type)
+    if ! status[:forces!] && (status[:energy_forces!] ||  status[:energy_forces_virial!])
+        tmp = generate_nonalloc_forces_from_energy_forces!(calc_type)
+        push!(out, tmp)
+    elseif ! status[:forces!] &&
+            (status[:forces] ||  status[:calculate_forces] || status[:energy_forces] ||  status[:energy_forces_virial])
+        tmp = generate_nonallocating_forces(calc_type)
+        push!(out, tmp)
+    end
+
+    # Generate only virial calls
+    if status[:virial] ||  status[:calculate_virial]
+        if status[:virial] && ! status[:calculate_virial]
+            tmp = generate_calculator_virial(calc_type)
+            push!(out, tmp)
+        elseif ! status[:virial] && status[:calculate_virial]
+            tmp = generate_virial(calc_type)
+            push!(out, tmp)
+        end
+    elseif status[:energy_forces_virial]
+        tmp = generate_virial_from_energy_forces_virial(calc_type)
+        push!(out, tmp)
+        tmp = generate_calculator_virial(calc_type)
+        push!(out, tmp)
+    elseif status[:energy_forces_virial!]
+        tmp = generate_virial_from_energy_forces_virial!(calc_type)
+        push!(out, tmp)
+        tmp = generate_calculator_virial(calc_type)
+        push!(out, tmp)
+    end
+
+    # Combination calls
+    if ! status[:energy_forces] && status[:energy_forces!]
+        tmp = generate_energy_forces_from_energy_forces!(calc_type)
+        push!(out, tmp)
+    elseif ! status[:energy_forces] && status[:energy_forces_virial]
+        tmp = generate_energy_forces_from_energy_forces_virial(calc_type)
+        push!(out, tmp)
+    elseif ! status[:energy_forces] && status[:energy_forces_virial!]
+        tmp = generate_energy_forces_from_energy_forces_virial!(calc_type)
+        push!(out, tmp)
+    end
+
+    if ! status[:energy_forces!] &&  status[:energy_forces_virial!]
+        tmp = generate_nonalloc_energy_forces_from_energy_forces_virial!(calc_type)
+        push!(out, tmp)
+    end
+
+    if ! status[:energy_forces_virial] && status[:energy_forces_virial!]
+        tmp = generate_energy_forces_virial_from_energy_forces_virial!(calc_type)
+        push!(out, tmp)
+
+    end
+
+    if ! status[:calculate_energy] && ! status[:calculate_forces] &&
+            (  status[:energy_forces] || status[:energy_forces!] ||
+            status[:energy_forces_virial] || status[:energy_forces!] || status[:energy_forces_virial!] )
+        # if this is true then optimized energy_forces exists
+        tmp = generate_calculate_energy_forces_from_energy_forces(calc_type)
+        push!(out, tmp)
+    end
+
+    if ! status[:calculate_energy] && ! status[:calculate_forces] && ! status[:calculate_virial] &&
+            ( status[:energy_forces_virial] || status[:energy_forces_virial!] )
+        # if this is true then optimized energy_forces_virial exists
+        tmp = generate_calculate_energy_forces_virial_from_energy_forces_virial(calc_type)
+        push!(out, tmp)
+    end
```

Comment:
This is a lot of code. Can this not be simplified by taking some flags (symbols) which the developer has to specify ? This makes this function shorter, easier to understand what it does and its action more explicit. I am almost sure this will avoid bugs in the long run.